src/opm/input/eclipse/Parser/ParserEnums.cpp

```diff
@@ -17,88 +17,95 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include <stdexcept>
-
 #include <opm/input/eclipse/Parser/ParserEnums.hpp>
 
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
 namespace Opm {
 
     /*****************************************************************/
 
-    const std::string ParserKeywordSizeEnum2String(ParserKeywordSizeEnum enumValue) {
-        switch (enumValue) {
-        case SLASH_TERMINATED:
-            return "SLASH_TERMINATED";
-            break;
-        case FIXED:
-            return "FIXED";
-            break;
-        case OTHER_KEYWORD_IN_DECK:
-            return "OTHER_KEYWORD_IN_DECK";
-            break;
-        case UNKNOWN:
-            return "UNKNOWN";
-            break;
-        case FIXED_CODE:
-            return "FIXED_CODE";
-            break;
-        case DOUBLE_SLASH_TERMINATED:
-            return "DOUBLE_SLASH_TERMINATED";
-            break;
-        default:
+    std::string ParserKeywordSizeEnum2String(ParserKeywordSizeEnum enumValue)
+    {
+        const auto enumMap = std::map<ParserKeywordSizeEnum, std::string> {
+            { ParserKeywordSizeEnum::SLASH_TERMINATED       , "SLASH_TERMINATED"        },
+            { ParserKeywordSizeEnum::FIXED                  , "FIXED"                   },
+            { ParserKeywordSizeEnum::OTHER_KEYWORD_IN_DECK  , "OTHER_KEYWORD_IN_DECK"   },
+            { ParserKeywordSizeEnum::UNKNOWN                , "UNKNOWN"                 },
+            { ParserKeywordSizeEnum::FIXED_CODE             , "FIXED_CODE"              },
+            { ParserKeywordSizeEnum::DOUBLE_SLASH_TERMINATED, "DOUBLE_SLASH_TERMINATED" },
+            { ParserKeywordSizeEnum::SPECIAL_CASE_ROCK      , "SPECIAL_CASE_ROCK"       },
+        };
```

Comment:
Seriously? Why?
IMHO
1. We create a map that will need to sort the entries of the initializer list
2. Then we do a binary search to find the entry

Seems like a lot of overhead and does not make the code much clearer.

I am not even sure whether I would recommend using this if you would be using `static constexpr std::map<ParserKeywordSizeEnum, std::string> ` instead (which would only increase compile time).

If you are that allergic to switch-case statements, then use a std::array for the lookup.

Reply:
Same holds for similar stuff below.

Reply:
I'm willing to take a performance hit here since these functions are mainly used while generating the C++ source code for the keywords, not while running the simulator or parser.  I considered marking the tables `static`, but that would keep them in memory for the duration of the process if something actually **does** call the functions at runtime.

As for "why", I addressed that in the commit message:
> Prefer Data to Logic
>
> Switch if/else chain out in favour of a map-based approach.  This simplifies adding new enumerators. [...]

The array-based approach only works for converting enumerators to the string value, not for the reverse direction.

Reply:
> The array-based approach only works for converting enumerators to the string value, not for the reverse direction.

Which seems to be the case here if I am not mistaking. The function takes an enum value and returns a string.

The problem is that in the future somebody might use this conversion in the simulator. There is nothing to prevent that.

Reply:
This would be my suggestion without switch-case:
````
const std::string[] strs = {"SLASH_TERMINATED", 
                                  "FIXED", 
                                  "OTHER_KEYWORD_IN_DECK",  
                                  "UNKNOWN" , 
                                  "FIXED_CODE" 
                                   "DOUBLE_SLASH_TERMINATED"
                                   "SPECIAL_CASE_ROCK"};
return strs[std::size_t(enumValue)];
```

>  > The array-based approach only works for converting enumerators to the string value, not for the reverse direction.

> Which seems to be the case here if I am not mistaking.

But not for the rest. Those are looking up an enum value by string indeed. I still neither like the new approach more than the original code nor see the improvement. But alas, there should be personal freedom to some extent :wink:. At least I appreciate that usually.

Nevertheless I would really appreciate it if you would change the enum-to-string lookup... What do you think?

Reply:
>  I would really appreciate it if you would change the enum-to-string lookup... What do you think?

I've pushed an update that switches to using a `std::array` instead.  I do think that's a more fragile option since it depends on

1. Keeping the two lists in the same order
2. Enumerator values being exactly 0, 1, ..., N-1

but it terms of lookup speed, it is obviously strictly better.

Reply:
Perfect. Thanks a lot.